File: python_src/cartridge_sdk/ui/chart.py
"""Chart widgets: SparkLine and LineChart."""

from __future__ import annotations

from typing import TYPE_CHECKING

import pygame

from cartridge_sdk.input import InputEvent
from cartridge_sdk.screen import _get_font
from cartridge_sdk.ui.widget import Widget

if TYPE_CHECKING:
    from cartridge_sdk.theme import Theme


class SparkLine(Widget):
    """Mini inline line chart for embedding in rows."""

    def __init__(self, data: list[float] | None = None, color: tuple | None = None) -> None:
        self.data: list[float] = data or []
        self.color = color

    def draw(self, surface: pygame.Surface, rect: pygame.Rect, theme: Theme) -> None:
        data = self.data
        if len(data) < 2:
            return

        color = self.color or theme.accent
        mn = min(data)
        mx = max(data)
        rng = mx - mn if mx != mn else 1.0

        points = []
        for i, v in enumerate(data):
            px = rect.x + int(i / (len(data) - 1) * (rect.width - 1))
            py = rect.y + rect.height - 1 - int(((v - mn) / rng) * (rect.height - 1))
            points.append((px, py))

        if len(points) >= 2:
            pygame.draw.lines(surface, color, False, points, 2)


class LineChart(Widget):
    """Full chart with axis labels and grid lines."""

    def __init__(
        self,
        data: list[float] | None = None,
        labels: list[str] | None = None,
        color: tuple | None = None,
        title: str = "",
    ) -> None:
        self.data: list[float] = data or []
        self.labels: list[str] = labels or []
        self.color = color
        self.title = title

    def draw(self, surface: pygame.Surface, rect: pygame.Rect, theme: Theme) -> None:
        data = self.data
        if len(data) < 2:
            font = _get_font("mono", 13)
            msg = font.render("No chart data", True, theme.text_dim)
            surface.blit(msg, (rect.x + rect.width // 2 - msg.get_width() // 2,
                               rect.y + rect.height // 2 - msg.get_height() // 2))
            return

        color = self.color or theme.accent
        font_label = _get_font("mono", 11)
        font_title = _get_font("mono_bold", 13)

        # Margins
        left_margin = 60
        right_margin = 12
        top_margin = 8
        bottom_margin = 24

        if self.title:
            title_surf = font_title.render(self.title, True, theme.text_dim)
            surface.blit(title_surf, (rect.x + left_margin, rect.y + 2))
            top_margin += font_title.get_linesize()

        chart_x = rect.x + left_margin
        chart_y = rect.y + top_margin
        chart_w = rect.width - left_margin - right_margin
        chart_h = rect.height - top_margin - bottom_margin

        if chart_w <= 0 or chart_h <= 0:
            return

        mn = min(data)
        mx = max(data)
        rng = mx - mn if mx != mn else 1.0

        # Grid lines and Y labels (5 lines)
        grid_color = (
            min(theme.border[0] + 8, 255),
            min(theme.border[1] + 8, 255),
            min(theme.border[2] + 8, 255),
        )
        num_grid = 4
        for i in range(num_grid + 1):
            gy = chart_y + int(i / num_grid * chart_h)
            pygame.draw.line(surface, grid_color, (chart_x, gy), (chart_x + chart_w, gy))

            val = mx - (i / num_grid) * rng
            if abs(val) >= 1000:
                label_text = f"${val:,.0f}"
            elif abs(val) >= 1:
                label_text = f"${val:.2f}"
            else:
                label_text = f"${val:.4f}"
            label_surf = font_label.render(label_text, True, theme.text_dim)
            surface.blit(label_surf, (rect.x + 2, gy - label_surf.get_height() // 2))

        # Chart line
        points = []
        for i, v in enumerate(data):
            px = chart_x + int(i / (len(data) - 1) * (chart_w - 1))
            py = chart_y + chart_h - 1 - int(((v - mn) / rng) * (chart_h - 1))
            points.append((px, py))

        # Fill area under the line (subtle)
        if len(points) >= 2:
            fill_points = list(points) + [
                (points[-1][0], chart_y + chart_h),
                (points[0][0], chart_y + chart_h),
            ]
            fill_color = (color[0], color[1], color[2], 25)
            fill_surface = pygame.Surface((chart_w, chart_h), pygame.SRCALPHA)
            shifted_points = [(p[0] - chart_x, p[1] - chart_y) for p in fill_points]
            if len(shifted_points) >= 3:
                pygame.draw.polygon(fill_surface, fill_color, shifted_points)
                surface.blit(fill_surface, (chart_x, chart_y))

            pygame.draw.lines(surface, color, False, points, 2)

            # End dot
            last = points[-1]
            pygame.draw.circle(surface, color, last, 3)

        # X-axis labels
        if self.labels:
            n_labels = min(5, len(self.labels))
            step = max(1, len(self.labels) // n_labels)
            for i in range(0, len(self.labels), step):
                lx = chart_x + int(i / max(1, len(self.labels) - 1) * (chart_w - 1))
                label_surf = font_label.render(self.labels[i], True, theme.text_dim)
                surface.blit(label_surf, (lx - label_surf.get_width() // 2,
                                          chart_y + chart_h + 4))

        # Border
        pygame.draw.rect(surface, grid_color, (chart_x, chart_y, chart_w, chart_h), 1)

File: crates/cartridge-core/src/ui/chart.rs
//! Chart widgets: SparkLine (mini inline) and LineChart (full with axes).

use sdl2::pixels::Color;
use sdl2::rect::Rect;

use crate::screen::Screen;

/// Mini inline sparkline chart for embedding in rows or small spaces.
///
/// Draws a simple polyline of data points scaled to fit `rect`.
pub struct SparkLine {
    pub data: Vec<f32>,
    pub color: Option<Color>,
}

impl SparkLine {
    pub fn new(data: Vec<f32>) -> Self {
        Self { data, color: None }
    }

    pub fn with_color(mut self, color: Color) -> Self {
        self.color = Some(color);
        self
    }

    /// Draw the sparkline within `rect`. Uses `screen.draw_sparkline` internally.
    pub fn draw(&self, screen: &mut Screen, rect: Rect) {
        if self.data.len() < 2 {
            return;
        }
        let baseline = Some(Color::RGB(
            screen.theme.border.r.saturating_add(8),
            screen.theme.border.g.saturating_add(8),
            screen.theme.border.b.saturating_add(8),
        ));
        screen.draw_sparkline(&self.data, rect, self.color, baseline);
    }
}

/// Full chart with Y-axis labels, grid lines, X-axis labels, and a data line.
///
/// Designed for financial data or system metrics. Draws grid, axis labels,
/// the data polyline, and an end-point dot.
pub struct LineChart {
    pub data: Vec<f32>,
    pub labels: Vec<String>,
    pub color: Option<Color>,
    pub title: String,
    /// Format function for Y-axis labels. Defaults to "$X.XX" style.
    /// Set to `None` for the default formatter.
    pub y_format: Option<fn(f32) -> String>,
}

impl LineChart {
    pub fn new(data: Vec<f32>) -> Self {
        Self {
            data,
            labels: Vec::new(),
            color: None,
            title: String::new(),
            y_format: None,
        }
    }

    pub fn with_labels(mut self, labels: Vec<String>) -> Self {
        self.labels = labels;
        self
    }

    pub fn with_color(mut self, color: Color) -> Self {
        self.color = Some(color);
        self
    }

    pub fn with_title(mut self, title: &str) -> Self {
        self.title = title.to_string();
        self
    }

    pub fn with_y_format(mut self, f: fn(f32) -> String) -> Self {
        self.y_format = Some(f);
        self
    }

    /// Draw the chart within `rect`.
    pub fn draw(&self, screen: &mut Screen, rect: Rect) {
        let theme = screen.theme;
        let data = &self.data;

        if data.len() < 2 {
            let text = "No chart data";
            let tw = screen.get_text_width(text, 13, false);
            let th = screen.get_line_height(13, false);
            screen.draw_text(
                text,
                rect.x() + (rect.width() as i32 - tw as i32) / 2,
                rect.y() + (rect.height() as i32 - th as i32) / 2,
                Some(theme.text_dim),
                13,
                false,
                None,
            );
            return;
        }

        let color = self.color.unwrap_or(theme.accent);

        // Margins
        let left_margin = 60i32;
        let right_margin = 12i32;
        let mut top_margin = 8i32;
        let bottom_margin = 24i32;

        // Title
        if !self.title.is_empty() {
            let title_lh = screen.get_line_height(13, true);
            screen.draw_text(
                &self.title,
                rect.x() + left_margin,
                rect.y() + 2,
                Some(theme.text_dim),
                13,
                true,
                None,
            );
            top_margin += title_lh as i32;
        }

        let chart_x = rect.x() + left_margin;
        let chart_y = rect.y() + top_margin;
        let chart_w = rect.width() as i32 - left_margin - right_margin;
        let chart_h = rect.height() as i32 - top_margin - bottom_margin;

        if chart_w <= 0 || chart_h <= 0 {
            return;
        }

        let mn = data.iter().copied().fold(f32::INFINITY, f32::min);
        let mx = data.iter().copied().fold(f32::NEG_INFINITY, f32::max);
        let rng = if (mx - mn).abs() > f32::EPSILON {
            mx - mn
        } else {
            1.0
        };

        // Grid lines and Y labels (5 horizontal lines)
        let grid_color = Color::RGB(
            theme.border.r.saturating_add(8),
            theme.border.g.saturating_add(8),
            theme.border.b.saturating_add(8),
        );
        let num_grid: i32 = 4;

        for i in 0..=num_grid {
            let gy = chart_y + (i as f32 / num_grid as f32 * chart_h as f32) as i32;
            screen.draw_line(
                (chart_x, gy),
                (chart_x + chart_w, gy),
                Some(grid_color),
                1,
            );

            let val = mx - (i as f32 / num_grid as f32) * rng;
            let label_text = if let Some(fmt) = self.y_format {
                fmt(val)
            } else {
                default_y_format(val)
            };

            let lh = screen.get_line_height(11, false);
            screen.draw_text(
                &label_text,
                rect.x() + 2,
                gy - lh as i32 / 2,
                Some(theme.text_dim),
                11,
                false,
                None,
            );
        }

        // Data points
        let points: Vec<(i32, i32)> = data
            .iter()
            .enumerate()
            .map(|(i, &v)| {
                let px =
                    chart_x + (i as f32 / (data.len() - 1) as f32 * (chart_w - 1) as f32) as i32;
                let py = chart_y + chart_h - 1
                    - (((v - mn) / rng) * (chart_h - 1) as f32) as i32;
                (px, py)
            })
            .collect();

        // Draw the line segments
        if points.len() >= 2 {
            for window in points.windows(2) {
                screen.draw_line(window[0], window[1], Some(color), 2);
            }

            // End dot
            let last = *points.last().unwrap();
            screen.draw_circle(last.0, last.1, 3, color);
        }

        // X-axis labels
        if !self.labels.is_empty() {
            let n_labels = 5.min(self.labels.len());
            let step = (self.labels.len() / n_labels).max(1);
            let mut i = 0;
            while i < self.labels.len() {
                let lx = chart_x
                    + (i as f32 / (self.labels.len() - 1).max(1) as f32 * (chart_w - 1) as f32)
                        as i32;
                let lw = screen.get_text_width(&self.labels[i], 11, false);
                screen.draw_text(
                    &self.labels[i],
                    lx - lw as i32 / 2,
                    chart_y + chart_h + 4,
                    Some(theme.text_dim),
                    11,
                    false,
                    None,
                );
                i += step;
            }
        }

        // Chart border
        screen.draw_rect(
            Rect::new(chart_x, chart_y, chart_w as u32, chart_h as u32),
            Some(grid_color),
            false,
            0,
            None,
        );
    }
}

fn default_y_format(val: f32) -> String {
    if val.abs() >= 1000.0 {
        // Rust doesn't have a built-in comma formatter; format without commas.
        format!("${:.0}", val)
    } else if val.abs() >= 1.0 {
        format!("${:.2}", val)
    } else {
        format!("${:.4}", val)
    }
}
